### PR TITLE
ipa-getkeytab enhancements

### PR DESCRIPTION
**Pull request by martbab**
This PR implements '-H' and '-Y' options mentioned in
https://fedorahosted.org/freeipa/ticket/6409 along with the ability to specify
CA cert on the command line (which proved useful during the work on installer
refactoring).

Since my C skills are not at the level I would like them to be it would be nice
if you point out even the tiniest mistakes, risky code or non-idiomatic usage.

Also the test case `test_retrieval_using_plain_ldap` fails due to unsuccesful
simple bind. I wanted to implement StartTLS for simple binds over ldap://, but
I get the following errors in dirsrv error log:

        [01/Nov/2016:10:44:52.395126000 +0000] connection - conn=883 fd=135
        Incoming BER Element was 3 bytes, max allowable is 209715200 bytes.
        Change the nsslapd-maxbersize attribute in cn=config to increase.

I guess there is something fishy with the way I initialize the StartTLS
session. I would appreciate your help with it.

### CODE REVIEW

File: client/ipa-getkeytab.c [hunk 1]
```diff
@@ -712,6 +717,8 @@ int main(int argc, const char *argv[])
               _("LDAP DN"), _("DN to bind as if not using kerberos") },
 	    { "bindpw", 'w', POPT_ARG_STRING, &bindpw, 0,
               _("LDAP password"), _("password to use if not using kerberos") },
+	    { "cacert", 'c', POPT_ARG_STRING, &ca_cert_file, 0,
+            _("Path to the IPA CA certificate"), _("IPA CA certificate")},
```
Comment (simo5) [inline client/ipa-getkeytab.c]: indentation does not follow the style of the rest of the code here

File: client/ipa-getkeytab.c [hunk 2]
```diff
@@ -137,26 +158,48 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
+                             char **ldap_uri)
 {
-	char* url = NULL;
-	int  url_len = snprintf(url,0,"%s://%s:%d",scheme,servername,port) +1;
+    char *url = NULL;
+    int url_len = 0;
+    int port = 0;
+    char *scheme = NULL;
+
+    if (!mech) {
```
Comment (simo5) [inline client/ipa-getkeytab.c]: I do not understand why you are using the mech to determine the URI scheme, these are unrelated things.

File: client/ipa-getkeytab.c [hunk 3]
```diff
@@ -137,26 +158,48 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
+        scheme = "ldaps";
+        port = 636;
+    } else if (strcmp(mech, LDAP_SASL_GSSAPI) == 0) {
+        scheme = "ldap";
+        port = 389;
+    }
+    else {
+        return 1;
+    }
+    url_len = snprintf(url,0,"%s://%s:%d",scheme,servername,port) +1;
```
Comment (simo5) [inline client/ipa-getkeytab.c]: please just use asprintf() in here.

File: client/ipa-getkeytab.c [hunk 4]
```diff
@@ -804,6 +890,14 @@ int main(int argc, const char *argv[])
             exit(2);
         }
     }
+    if (server) {
+        ret = ipa_server_to_uri(server, sasl_mech, &ldap_uri);
+        if (ret) {
+            fprintf(stderr, _("Incompatible combination of server name and "
+                              "SASL bind mechanism\n"));
```
Comment (simo5) [inline client/ipa-getkeytab.c]: This error message looks wrong.
ipa_server_to_uri() currently fails only if the mechanism is not recognized (but that one is checked earlier already), or if string allocation fails.

File: client/ipa-getkeytab.c [hunk 5]
```diff
@@ -837,7 +931,7 @@ int main(int argc, const char *argv[])
 		exit(4);
 	}
 
-	if (NULL == bindpw) {
+	if (NULL == bindpw && strncmp(sasl_mech, LDAP_SASL_GSSAPI, 79) == 0) {
```
Comment (simo5) [inline client/ipa-getkeytab.c]: 79 ?

File: client/ipa-getkeytab.c [hunk 6]
```diff
@@ -196,6 +196,21 @@ static int ipa_ldap_init(LDAP **ld, const char *ldap_uri)
 	return rc;
 }
 
+static int ipa_starttls_init(LDAP *ld, const char *ldap_uri)
+{
+    int rc = LDAP_SUCCESS;
+    char *startswith_ldap = NULL;
+
+    startswith_ldap = strstr("ldap://", ldap_uri);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: ldap:// can be only at the start of the string, pleae use strncmp(ldap_uri, "ldap://", 7)

File: client/ipa-getkeytab.c (hunk 2 above)
Comment (martbab) [inline client/ipa-getkeytab.c]: That is probably some leftover from previous incarnations of the patch.  If I manage to make StartTLS work this may actually be irrelevant since we may use ldap://$FQDN and have secure connection.

**Comment by martbab** [comment]
Thank you for review @simo5 . I have fixed the issues and reworked the LDAP initialization and binding logic a bit to clean it up. It produced green tests for me. I have also updated the command man page as I missed that during initial work.

**Comment by martbab** [comment]
Bump for review, this is prerequisite to further installer refactoring work.

File: client/ipa-getkeytab.c [hunk 7]
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
+    char *scheme = "ldap";
 
-	url = (char *)malloc (url_len);
-	if (!url){
-		fprintf(stderr, _("Out of memory \n"));
-		return LDAP_NO_MEMORY;
-	}
-	sprintf(url,"%s://%s:%d",scheme,servername,port);
-	int rc = ldap_initialize(ld, url);
+    url_len = asprintf(&url,"%s://%s:%d",scheme,servername,port);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: space after comma

File: client/ipa-getkeytab.c [hunk 8]
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
 
-	free(url);
-	return rc;
+    if (url_len == -1){
```
Comment (simo5) [inline client/ipa-getkeytab.c]: space after close parens

File: client/ipa-getkeytab.c [hunk 9]
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
+    return rc;
+}
+
+static int ipa_tls_ssl_init(LDAP *ld, const char *ldap_uri)
+{
+    int ret = LDAP_SUCCESS;
+    int tls_hard = LDAP_OPT_X_TLS_HARD;
+    int tls_demand = LDAP_OPT_X_TLS_DEMAND;
+    char *ldap = "ldap://";
+    char *ldaps = "ldaps://";
```
Comment (simo5) [inline client/ipa-getkeytab.c]: there is no need to use an actual variable, beside these would have to be defined as const char as the strings are constant.

Use defines:
```#define SCHEMA_LDAP "ldap://"```
```#define SCHEMA_LDAPS "ldaps://"```
Then for lenght you can use sizeof(SCHEMA_LDAP) -1

File: client/ipa-getkeytab.c [hunk 10]
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
             fprintf(stderr, _("Unable to set LDAP_OPT_X_TLS\n"));
-            goto done;
-        }
-    } else {
-        ret = ipa_ldap_init(&ld, "ldap", server_name, 389);
-        if (ret != LDAP_SUCCESS) {
-            fprintf(stderr, _("Unable to init for ldap(389) connection\n"));
             return ret;
         }
+
```
Comment (simo5) [inline client/ipa-getkeytab.c]: don't add empty line

File: client/ipa-getkeytab.c [hunk 11]
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
                                &bv, NULL, NULL, NULL);
         if (ret != LDAP_SUCCESS) {
             fprintf(stderr, _("Simple bind failed\n"));
-            goto done;
```
Comment (simo5) [inline client/ipa-getkeytab.c]: This looks wrong, if bind fails we should get out

File: client/ipa-getkeytab.c [hunk 12]
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
...
+
         if (ret != LDAP_SUCCESS) {
 #ifdef LDAP_OPT_DIAGNOSTIC_MESSAGE
             ldap_get_option(ld, LDAP_OPT_DIAGNOSTIC_MESSAGE, (void*)&msg);
 #endif
             fprintf(stderr, "SASL Bind failed %s (%d) %s!\n",
                             ldap_err2string(ret), ret, msg ? msg : "");
             goto done;
+            }
         }
```
Comment (simo5) [inline client/ipa-getkeytab.c]: please make sure the close parens matches the same indentation of the open one at all times

File: client/ipa-getkeytab.c [hunk 13]
```diff
@@ -221,21 +274,26 @@ static int ipa_ldap_bind(const char *server_name, krb5_principal bind_princ,
...
             fprintf(stderr, _("Simple bind failed\n"));
-            goto done;
         }
     } else {
-        ret = ldap_sasl_interactive_bind_s(ld, NULL, "GSSAPI",
-                                           NULL, NULL, LDAP_SASL_QUIET,
-                                           ldap_sasl_interact, bind_princ);
+        if (strcmp(mech, LDAP_SASL_EXTERNAL) == 0) {
+            ret = ldap_sasl_bind_s(ld, NULL, LDAP_SASL_EXTERNAL,
+                               NULL, NULL, NULL, NULL);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: bad indent, sshould start under 'ld'

File: client/ipa-getkeytab.c [hunk 14]
```diff
@@ -783,7 +857,34 @@ int main(int argc, const char *argv[])
 		exit(10);
 	}
 
-    if (!server) {
+    if (NULL != binddn && NULL != sasl_mech) {
+        fprintf(stderr, _("Cannot specify both SASL mechanism "
+                          "and bind DN simultaneously.\n"));
+		if (!quiet)
+			poptPrintUsage(pc, stderr, 0);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: incorrect indentation

File: client/ipa-getkeytab.c [hunk 15]
```diff
@@ -783,7 +857,34 @@ int main(int argc, const char *argv[])
...
+
+    if (!binddn && !sasl_mech) {
+        sasl_mech = LDAP_SASL_GSSAPI;
+    }
+
+    if (server && ldap_uri) {
+        fprintf(stderr, _("Cannot specify server and LDAP uri "
+                          "simultaneously.\n"));
+		if (!quiet)
+			poptPrintUsage(pc, stderr, 0);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: incorrect indentation

File: client/man/ipa-getkeytab.1 [hunk 16]
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
 The LDAP DN to bind as when retrieving a keytab without Kerberos credentials. Generally used with the \fB\-w\fR option.
 .TP
 \fB\-w, \-\-bindpw\fR
-The LDAP password to use when not binding with Kerberos.
+The LDAP password to use when not binding with Kerberos. \fB\-D\fR and \fB\-w\fR can not be used together with \fB\-Y\fR.
+.TP
+\fB\-\-cacert\fR
+The path to IPA CA certificate used to validate LDAPS/STARTTLS connection.
```
Comment (simo5) [inline client/man/ipa-getkeytab.1]: -> path to 'the' IPA CA certificate
-> connections (plural)

File: client/man/ipa-getkeytab.1 [hunk 17]
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
...
+Defaults to /etc/ipa/ca.crt
+.TP
+\fB\-H, \-\-ldapuri\fR
+LDAP URI. If ldap:// is specified, STARTTLS session is initiated by default. Can not be used with \fB\-s\fR.
```
Comment (simo5) [inline client/man/ipa-getkeytab.1]: remove 'session'

File: client/man/ipa-getkeytab.1 [hunk 18]
```diff
@@ -96,7 +96,17 @@ Use this password for the key instead of one randomly generated.
...
+.TP
+\fB\-Y, \-\-mech\fR
+SASL mechanism to use if fB\-D\fR and \fB\-w\fR are not used. Choose between GSSAPI and EXTERNAL.
```
Comment (simo5) [inline client/man/ipa-getkeytab.1]: -> s/Choose between GSSAPI and/Use either GSSAPI or/

File: client/ipa-getkeytab.c [hunk 19]
```diff
@@ -135,61 +158,86 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
         if (ret != LDAP_SUCCESS) {
-            fprintf(stderr, _("Unable to init for ldaps(636) connection\n"));
+            fprintf(stderr, _("Unable to initialize STARTTLS session\n"));
             return ret;
         }
-
-        ssl = LDAP_OPT_X_TLS_HARD;;
-        ret = ldap_set_option(ld, LDAP_OPT_X_TLS, &ssl);
+    }
+    else if (strncmp(ldap_uri, ldaps, strlen(ldaps)) == 0) {
```
Comment (simo5) [inline client/ipa-getkeytab.c]: } else if,
don't go to a new line wth the else

**Comment by martbab** [comment]
Another bump for review.

File: client/ipa-getkeytab.c [hunk 20]
```diff
@@ -135,62 +161,82 @@ int filter_keys(krb5_context krbctx, struct keys_container *keys,
...
+    int port = 389;
 
-	url = (char *)malloc (url_len);
-	if (!url){
-		fprintf(stderr, _("Out of memory \n"));
-		return LDAP_NO_MEMORY;
-	}
-	sprintf(url,"%s://%s:%d",scheme,servername,port);
-	int rc = ldap_initialize(ld, url);
+    url_len = asprintf(&url,"%s%s:%d", SCHEMA_LDAP, servername, port);
```
Comment (simo5) [inline client/ipa-getkeytab.c]: space after comma

**Comment by martbab** [comment]
Done.

**Comment by martbab** [comment]
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/294fc3dc5645eeb7942908c3e351c06aa0af329e
https://fedorahosted.org/freeipa/changeset/0c68c27e51c2a30265a760382d7d4fab7d21937b
https://fedorahosted.org/freeipa/changeset/8480d0e3333f6813439e7b3321a0e33ce80d30f1
https://fedorahosted.org/freeipa/changeset/2725e440bf1e4930f9b1d19223424bcb0d4b7066
